File: 20/aoc10.cc
#include <algorithm>
#include <iostream>
#include <fstream>
#include <map>
#include <string>
#include <vector>
#include "aoc_helper.h"

int joltageProduct(std::vector<long> data) {
    int diffOne = 0;
    int diffThree = 0;

    std::sort(data.begin(), data.end());
    
    // initial diff
    int initialDiff = data[0] - 0;
    if ( initialDiff == 1) {
        diffOne++;
    } else if ( initialDiff == 3) {
        diffThree++;
    }

    for (int i = 0; i < data.size() - 1; i++) {
        int curr = data[i];
        int next = data[i + 1];
        int diff = next - curr;
        if (diff == 1) {
            diffOne++;
        } else if (diff == 3) {
            diffThree++;
        }
    }
    // device's built-in adapter 
    diffThree++;
    return diffThree * diffOne;
}


int main() {
    //read input from file
    std::vector<long> data;
    data = getInputLong("./input/aoc10.txt");

    int joltage = joltageProduct(data);
    std::cout << "the joltage product is " << joltage << std::endl;
}

File: 20/python/aoc10.py

def joltage(data):
    data.sort()
    one_j, two_j, three_j = 0, 0, 0
    i = 0
    while i < len(data):
        current_j = data[i]
        if i == 0:
            prev_j = 0
        else:
            prev_j = data[i-1]

        diff = current_j - prev_j
        if diff == 3:
            three_j += 1
        elif diff == 2:
            two_j += 1
        elif diff == 1:
            one_j += 1
        i += 1
    # device's built in adapter
    three_j += 1
    return three_j * one_j

def calculate_combos(el, map):
    combos = 0
    diff_one, diff_two, diff_three = el - 1, el - 2, el - 3
    if diff_one in map:
        combos += map[diff_one]
    if diff_two in map:
        combos += map[diff_two]
    if diff_three in map:
        combos += map[diff_three]
    map[el] = combos
    return map

def joltage2(data):
    data.sort()
    last_el = max(data) + 3
    data.append(last_el)
    combos = {0: 1}
    for el in data:
        combos = calculate_combos(el, combos)
    
    return combos[last_el]

if __name__ == "__main__":
    text_file = open("../input/aoc10.txt", "r")
    string_data = text_file.readlines()
    data = list(map(lambda x: int(x), string_data))
    ans = joltage2(data)
    print("the total number of distinct ways to arrange the adapters is {}".format(ans))
